lib/evaluate/check-call.cc

```diff
@@ -0,0 +1,199 @@
+#include "../parser/message.h"
+#include <map>
+#include <string>
+
+using namespace Fortran::parser::literals;
+
+namespace Fortran::evaluate {
+
+static void CheckImplicitInterfaceArg(
+    ActualArgument &arg, parser::ContextualMessages &messages) {
+  if (const auto &kw{arg.keyword}) {
+    messages.Say(*kw,
+        "Keyword '%s=' cannot appear in a reference to a procedure with an implicit interface"_err_en_US,
+        *kw);
+  }
+  if (auto type{arg.GetType()}) {
+    if (type->IsAssumedType()) {
+      messages.Say(
+          "Assumed type argument requires an explicit interface"_err_en_US);
+    } else if (type->IsPolymorphic()) {
+      messages.Say(
+          "Polymorphic argument requires an explicit interface"_err_en_US);
+    } else if (type->category() == TypeCategory::Derived) {
+      auto &derived{type->GetDerivedTypeSpec()};
+      if (!derived.parameters().empty()) {
+        messages.Say(
+            "Parameterized derived type argument requires an explicit interface"_err_en_US);
+      }
+    }
+  }
+  if (const auto *expr{arg.UnwrapExpr()}) {
+    if (auto named{ExtractNamedEntity(*expr)}) {
+      const semantics::Symbol &symbol{named->GetLastSymbol()};
+      if (symbol.Corank() > 0) {
+        messages.Say(
+            "Coarray argument requires an explicit interface"_err_en_US);
+      }
+      if (const auto *details{
+              symbol.detailsIf<semantics::ObjectEntityDetails>()}) {
+        if (details->IsAssumedRank()) {
+          messages.Say(
+              "Assumed rank argument requires an explicit interface"_err_en_US);
+        }
+      }
+      if (symbol.attrs().test(semantics::Attr::ASYNCHRONOUS)) {
+        messages.Say(
+            "ASYNCHRONOUS argument requires an explicit interface"_err_en_US);
+      }
+      if (symbol.attrs().test(semantics::Attr::VOLATILE)) {
+        messages.Say(
+            "VOLATILE argument requires an explicit interface"_err_en_US);
+      }
+    }
+  }
+}
+
+static void CheckExplicitInterfaceArg(const ActualArgument &arg,
+    const characteristics::DummyArgument &dummy, FoldingContext &context) {
+  std::visit(
+      common::visitors{
+          [&](const characteristics::DummyDataObject &object) {
+            if (const auto *expr{arg.UnwrapExpr()}) {
+              if (auto type{characteristics::GetTypeAndShape(*expr, context)}) {
+                object.type.IsCompatibleWith(context.messages(), *type);
+              } else {
+                // TODO
+              }
+            } else {
+              // TODO
+            }
+          },
+          [&](const characteristics::DummyProcedure &) {
+            // TODO check effective procedure compatibility
+          },
+          [&](const characteristics::AlternateReturn &) {
+            // TODO check alternate return
+          },
+      },
+      dummy.u);
+}
+
+static void RearrangeArguments(const characteristics::Procedure &proc,
+    ActualArguments &arg, parser::ContextualMessages &messages) {
```

Comment:
`args` would be a better name than `arg`.

Reply:
Ok.